---
title: "Fine mapping genetic associations"
output: 
  html_document: 
    keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, echo=FALSE, warning=F, message=F}
suppressMessages(library("dplyr"))
suppressMessages(library("tidyr"))
suppressMessages(library("ggplot2"))
suppressMessages(library("susieR"))
suppressMessages(library("SummarizedExperiment"))
```

# Import data

Import CEDAR gene expression data SummarizedExperiment object

```{r}
se = readRDS("CEDAR_SummarizedExeperiment.rds")
se
```

Import genotype data near two genes (ARHGEF3 and LCT)

```{r}
arhgef3_genotypes = readRDS("CEDAR_ARHGEF3_genotypes.rds")
lct_genotyes = readRDS("CEDAR_LCT_genotypes.rds")
```

# Perform fine mapping for the ARHGEF3 gene

The CEDAR dataset gene expression data from 6 blood cell types (B-cells, monocytes, neutrophils, platelets and T-cells (CD4+ and CD8+)) and three intestinal tissues (ileum, transverse colon, rectum):

```{r}
table(se$qtl_group)
```

For the ARHGEF3 example, we first need to filter the SummarizedExperiment object to keep only the platelet samples.
```{r}
filtered_se = se[,se$qtl_group == "platelet"]
filtered_se
```

## Extracting the gene expression vector

Since CEDAR used Illumina HumanHT12_V4 microarray to measure gene expression (instead of more modern RNA-seq), we need to find out, which probe on microarray corresponded to the ARHGEF3 gene.

```{r}
#Find probe IDs for the CD14 gene
gene_metadata = rowData(filtered_se) %>% as.data.frame() %>% as.tbl() %>% dplyr::filter(gene_name == "ARHGEF3")
gene_metadata
```
Next we can extract this probe from the expression matrix

```{r}
#Extract gene expression matrix
expression_matrix = assays(filtered_se)$counts
expression_matrix[1:5, 1:5]
```

Replace sample_ids on columns with genotype ids to match the genotype matrix (We can do this easily, because in a SummarizedExperiment obejct, the the assays (e.g. expression matrix) as well as rowData and colData data frames are always sorted accoring to the row and colum ids)

```{r}
genotype_ids = colData(filtered_se) %>% as.data.frame() %>% dplyr::select(sample_id, genotype_id)
colnames(expression_matrix) = genotype_ids$genotype_id
expression_matrix[1:5, 1:5]
```

Extract the probe corresponding to ARHGEF3
```{r}
expression_vector = expression_matrix["ILMN_1781010",]
expression_vector[1:5]
```

## Extract corresponding individuals from the genotype data

```{r}
genotype_matrix = arhgef3_genotypes$genotypes[,names(expression_vector)]
standard_genotypes = t(genotype_matrix - apply(genotype_matrix, 1, mean))
standard_genotypes[1:5,1:5]
```

## Perform finemapping 

Run the SuSiE fine mapping model. Note that you can use the coverage paramter to change to coverage of the credible set. Increasing the coverage generally makes the credible set larger, but at the same time, reduces to probability that the true causal variant is accidentally excluded from the credible set. Coverage of 99% would mean that the causal variant should be included in the credible set with 99% probability.

```{r}
fitted <- susieR::susie(standard_genotypes, expression_vector,
                        verbose = TRUE,
                        compute_univariate_zscore = TRUE,
                        coverage = 0.95)
fitted$variant_id = colnames(standard_genotypes)
```

## Extract the variants belonging to the credible set together with their PIP
```{r}
credible_set = fitted$pip[fitted$sets$cs$L1]
names(credible_set) = fitted$variant_id[fitted$sets$cs$L1]
credible_set
```

The easiest way to to find out the rsid of this variant is to query the GTEx website with _b38 appended to the variant coordinates:
https://gtexportal.org/home/snp/chr3_56815721_T_C_b38

## Visualise the fine mapping results

Exract SNP positions for the variants included in fine mapping:

```{r}
pos_tbl = dplyr::tibble(snpid = colnames(standard_genotypes)) %>% dplyr::left_join(arhgef3_genotypes$snpspos)
```

Visualise fine mapping results on top of -log10 p-values z-scores from univariate analysis:

```{r}
susieR::susie_plot(fitted, y = "z")
```

Visualise posterior inclusion probability (PIP) for each variant directly

```{r}
susieR::susie_plot(fitted, y = "PIP")
```


